<a href="https://colab.research.google.com/github/username1103/tensorflow_example/blob/master/conv2D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Convolution**

1.   필터는 3차원 형태로 된 가중치의 모음
2.   필터 하나는 앞선 레이어의 결과인 '특징맵' 전체를 본다.
3.   필터들 개수 만큼 특징맵을 만든다.


#### 필터를 몇 개 사용할 것인가

#### 필터의 사이즈를 얼마로 할 것인가


In [3]:
import tensorflow as tf
import pandas as pd

# 데이터 로드
(독립, 종속), _ = tf.keras.datasets.mnist.load_data()
독립 = 독립.reshape(60000,28,28,1) # 흑백이미지도 3차원 형태로 변환해 준다고 생각
종속 = pd.get_dummies(종속)
print(독립.shape, 종속.shape)

11493376/11490434 [==============================] - 0s 0us/step
(60000, 28, 28, 1) (60000, 10)


In [4]:
# 모델 생성
X = tf.keras.layers.Input(shape=[28,28,1])
H = tf.keras.layers.Conv2D(3, kernel_size=5, activation='swish')(X) # 3개의 feature map 생성 
H = tf.keras.layers.Conv2D(6, kernel_size=5, activation='swish')(H) # 6개의 feature map 생성 kernel_size는 필터의 사이즈를 의미 
H = tf.keras.layers.Flatten()(H) # 표로 만든다 
H = tf.keras.layers.Dense(84, activation='swish')(H)
Y = tf.keras.layers.Dense(10, activation='softmax')(H)
model = tf.keras.models.Model(X,Y)
model.compile(loss='categorical_crossentropy', metrics='accuracy')

In [6]:
# 모델 학습
model.fit(독립, 종속, epochs=10)

1875/1875 [==============================] - 38s 3ms/step - loss: 1.1031 - accuracy: 0.8960
Epoch 2/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0833 - accuracy: 0.9762
Epoch 3/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0548 - accuracy: 0.9845
Epoch 4/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0384 - accuracy: 0.9895
Epoch 5/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0291 - accuracy: 0.9920
Epoch 6/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0266 - accuracy: 0.9922
Epoch 7/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0207 - accuracy: 0.9940
Epoch 8/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0181 - accuracy: 0.9946
Epoch 9/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0149 - accuracy: 0.9958
Epoch 10/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0162 - accura

In [9]:
# 모델 이용
pred = model.predict(독립[0:5])
pd.DataFrame(pred).round(2)

0    1    2    3    4    5    6    7    8    9
0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
1  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
2  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0
3  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0

In [11]:
# 비교하기
종속[0:5]

0  1  2  3  4  5  6  7  8  9
0  0  0  0  0  0  1  0  0  0  0
1  1  0  0  0  0  0  0  0  0  0
2  0  0  0  0  1  0  0  0  0  0
3  0  1  0  0  0  0  0  0  0  0
4  0  0  0  0  0  0  0  0  0  1

In [12]:
# 모델 확인
model.summary()

Model: "model"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         [(None, 28, 28, 1)]       0         
_________________________________________________________________
conv2d (Conv2D)              (None, 24, 24, 3)         78        
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 20, 20, 6)         456       
_________________________________________________________________
flatten (Flatten)            (None, 2400)              0         
_________________________________________________________________
dense (Dense)                (None, 84)                201684    
_________________________________________________________________
dense_1 (Dense)              (None, 10)                850       
Total params: 203,068
Trainable params: 203,068
Non-trainable params: 0
_______________________________________________________

[28,28,1] 의 이미지가 들어가서 [5,5,1]사이즈 필터 3개로 컨볼루션을 통해 [24,24,3]으로 변경
<p>=> 따라서 25개의 가중치가 총 3개 + 편항 3개 = 78개의 param</p>

[24,24,3] 의 이미지들을 [5,5,3] 사이즈 필터 6개로 컨볼루션을 통해 [20,20,6]으로 변경
<p>=> 따라서 채널이 3개인 25개의 가중치 필터가 6개 + 6개의 편향 = 456개의 param</p>

[20,20,6] 으로 변경된 이미지를 Flatten layer를 통해 표로 변경
<p>=> 20 * 20 * 6 = 2400</p>

2400개의 가중치 + 1개의 편향으로 하나의 값 생성 * 84개의 노드를 가진 히든레이어
<p>=> 2401 * 84 = 201684 </p>

84개의 가중치 * 1개의 편향으로 하나의 값 생성 * 10
<p>=> 85 * 10 = 850</p>